src/Package/Impl/ProjectSystem/Commands/SourceFilesCommand.cs

```diff
@@ -21,34 +29,76 @@
-                return true;
+                IEnumerable<string> rFiles = Enumerable.Empty<string>();
+
+                IFileSystem fs = new FileSystem();
+                var workflow = _interactiveWorkflowProvider.GetOrCreate();
+                try {
+                    var session = workflow.RSession;
+                    if (session.IsRemote) {
+                        var properties = _configuredProject.Services.ExportProvider.GetExportedValue<ProjectProperties>();
+                        
+                        string projectDir = Path.GetDirectoryName(_configuredProject.UnconfiguredProject.FullPath);
+                        string projectName = properties.GetProjectName();
+                        string remotePath = await properties.GetRemoteProjectPathAsync();
+
+                        var files = nodes.GetSelectedNodesPaths().Where(x =>
+                                   Path.GetExtension(x).EqualsIgnoreCase(RContentTypeDefinition.FileExtension) &&
+                                   fs.FileExists(x));
+                        workflow.ActiveWindow.InteractiveWindow.WriteLine(Resources.Info_CompressingFiles);
+                        
+                        string compressedFilePath = fs.CompressFiles(files, projectDir, new Progress<string>((p) => {
+                            workflow.ActiveWindow.InteractiveWindow.WriteLine(string.Format(Resources.Info_LocalFilePath, p));
+                            string dest = p.MakeRelativePath(projectDir).ProjectRelativePathToRemoteProjectPath(remotePath, projectName);
+                            workflow.ActiveWindow.InteractiveWindow.WriteLine(string.Format(Resources.Info_RemoteDestination, dest));
+
+                        }), CancellationToken.None);
+
+                        using (var fts = new DataTransferSession(session, fs)) {
+                            workflow.ActiveWindow.InteractiveWindow.WriteLine(Resources.Info_TransferingFiles);
+                            var remoteFile = await fts.SendFileAsync(compressedFilePath);
+                            await session.EvaluateAsync<string>($"rtvs:::save_to_project_folder({remoteFile.Id}, {projectName.ToRStringLiteral()}, '{remotePath.ToRPath()}')", REvaluationKind.Normal);
+                            workflow.ActiveWindow.InteractiveWindow.WriteLine(Resources.Info_TransferingFilesDone);
+                        }
+
+                        rFiles = files.Select(p => p.MakeRelativePath(projectDir).ProjectRelativePathToRemoteProjectPath(remotePath, projectName));
+                    } else {
+                        rFiles = nodes.GetSelectedNodesPaths().Where(x =>
+                                   Path.GetExtension(x).EqualsIgnoreCase(RContentTypeDefinition.FileExtension) &&
+                                   fs.FileExists(x));
+                    }
+
+                    workflow.Operations.SourceFiles(rFiles, echo);
+                    return true;
+                } catch (RHostDisconnectedException) {
```

Comment:
Are there more exceptions to catch, like IOException? With all the file handling that's done, seems likely we'll get one of those at some point.